design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,191 @@
+ formats KSQL won’t have any issues since the schema field names already are compatible with the current
+ KSQL naming requirements, in handling content with JSON format, KSQL would encounter limitations.
+ In JSON format, field names don’t necessarily conform to KSQL naming requirements and a valid JSON
+ message can have field names that cannot be accepted by KSQL. In addition to the above requirement,
+ the field names in KSQL should also comply with the identifier naming requirements in SQL in order to
+ have the statements parse correctly.
+
+In this KLIP, we propose to decouple the internal schema, that KSQL uses for query processing, from
+ the external schema that user content conforms to. The internal KSQL schema will be fully compliant
+ with KSQL requirements and the engine can use it to process queries with no issues. On the other
+ hand, external schema won’t be restricted with KSQL requirements and users can use any content with
+ any naming convention in KSQL.
+
+Users will be able to use quoted identifiers to declare and access fields that do not conform to
+the KSQL identifier requirements. KSQL will generate an internal schema with a one to one mapping
+from the external schema fields into the internal schema fields. All of the computation will use the
+field names from the internal schema.
+
+As an example, consider we want to declare s stream on a topic with values in JSON format. The following
+is a sample message value in this topic:
+
+```json
+{"@ID": 0, "@NAME": "foo", "@AMOUNT": 0.0}
+```
+
+As you can see, two of the field names start with “@” and the third one includes “.” in the field name.
+ Currently we cannot declare a stream or table on this topic since its content does not conform to KQSL
+ field naming requirements. However, we should be able to use quoted identifiers, which is supported by
+ our parser, to declare a stream or table on this topic and write queries over the declared stream or
+ table. Let’s assume we declare a stream on this topic, named foo. The following would be the DDL
+  statement:
+
+```sql
+CREATE STREAM foo
+  (“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
+WITH
+  (KAFKA_TOPIC=’foo’, VALUE_FORMAT = ‘JSON’);
+```
+
+KSQL will add the above stream with the declared schema to its metastore. Now, let’s assume we want
+to create a new stream from foo using the following CSAS statement:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+KSQL will generate an internal schema for foo before processing the above query and rewrites the query
+based on the internal schema. It then will use the rewritten query to process the messages from the topic.
+
+## Scope
+
+* Support for Quoted Identifiers in DDL and DML statements
+* Creation of internal schema for external schemas
+* Query rewrite to use the internal schemas
+* Query processing using the internal schemas
```

Comment:
* Quoting a field allows it to be a reserved work, right? Is that in scope?

Reply:
Yes, quoting a fields allows you it to have any string value in the quotes. You can have reserved words in the quotes and handling it is in the scope of this KLIP. For instance we should be able to handle the following:
```sql
SELECT "SELECT" from foo;
```

Reply:
So can we update the KLIP to include this in the scope?